pkg/loghttp/push/otlp.go

```diff
@@ -0,0 +1,373 @@
+			for k := 0; k < logs.Len(); k++ {
+				log := logs.At(k)
+
+				entry := otlpLogToPushEntry(log)
+				entry.StructuredMetadata = append(entry.StructuredMetadata, resourceAttributesAsStructuredMetadata...)
+				entry.StructuredMetadata = append(entry.StructuredMetadata, scopeAttributesAsStructuredMetadata...)
+				stream := pushRequestsByStream[labelsStr]
+				stream.Entries = append(stream.Entries, entry)
+				pushRequestsByStream[labelsStr] = stream
+
+				stats.structuredMetadataBytes[tenantsRetention.RetentionPeriodFor(userID, lbs)] += int64(labelsSize(entry.StructuredMetadata) - resourceAttributesAsStructuredMetadataSize - scopeAttributesAsStructuredMetadataSize)
+				stats.logLinesBytes[tenantsRetention.RetentionPeriodFor(userID, lbs)] += int64(len(entry.Line))
+				stats.numLines++
+				if entry.Timestamp.After(stats.mostRecentEntryTimestamp) {
+					stats.mostRecentEntryTimestamp = entry.Timestamp
+				}
+			}
+		}
+	}
+
+	pr := &push.PushRequest{
+		Streams: make([]push.Stream, 0, len(pushRequestsByStream)),
+	}
+
+	for _, stream := range pushRequestsByStream {
+		pr.Streams = append(pr.Streams, stream)
+	}
+
+	return pr
+}
+
+// otlpLogToPushEntry converts an OTLP log record to a Loki push.Entry.
+func otlpLogToPushEntry(log plog.LogRecord) push.Entry {
+	// copy log attributes and all the fields from log(except log.Body) to structured metadata
+	structuredMetadata := attributesToLabels(log.Attributes(), "")
+
+	// if log.Timestamp() is 0, we would have already stored log.ObservedTimestamp as log timestamp so no need to store again in structured metadata
+	if log.Timestamp() != 0 && log.ObservedTimestamp() != 0 {
+		structuredMetadata = append(structuredMetadata, push.LabelAdapter{
+			Name:  "observed_timestamp",
+			Value: fmt.Sprintf("%d", log.ObservedTimestamp().AsTime().UnixNano()),
+		})
+	}
+
+	if severityNum := log.SeverityNumber(); severityNum != plog.SeverityNumberUnspecified {
+		structuredMetadata = append(structuredMetadata, push.LabelAdapter{
+			Name:  "severity_number",
+			Value: fmt.Sprintf("%d", severityNum),
+		})
+	}
+	if severityText := log.SeverityText(); severityText != "" {
+		structuredMetadata = append(structuredMetadata, push.LabelAdapter{
+			Name:  "severity_text",
+			Value: severityText,
+		})
+	}
+
+	if droppedAttributesCount := log.DroppedAttributesCount(); droppedAttributesCount != 0 {
+		structuredMetadata = append(structuredMetadata, push.LabelAdapter{
+			Name:  "dropped_attributes_count",
+			Value: fmt.Sprintf("%d", droppedAttributesCount),
+		})
+	}
+	if logRecordFlags := log.Flags(); logRecordFlags != 0 {
+		structuredMetadata = append(structuredMetadata, push.LabelAdapter{
+			Name:  "flags",
+			Value: fmt.Sprintf("%d", logRecordFlags),
+		})
+	}
+
+	if traceID := log.TraceID(); !traceID.IsEmpty() {
+		structuredMetadata = append(structuredMetadata, push.LabelAdapter{
+			Name:  "trace_id",
+			Value: hex.EncodeToString(traceID[:]),
```

Comment:
We can call `traceID.String()` here

Reply:
It is not ideal to call `String()` as recommended by [otel lib](https://github.com/open-telemetry/opentelemetry-collector/blob/287b98f6973fd6baa278150b9fca8c83abea0af4/pdata/pcommon/traceid.go#L22-L26).